homeassistant/components/snmp/sensor.py

```diff
@@ -106,6 +108,17 @@ async def async_setup_platform(
     default_value = config.get(CONF_DEFAULT_VALUE)
     unique_id = config.get(CONF_UNIQUE_ID)
 
+    try:
+        # Try IPv4 first.
+        target = UdpTransportTarget((host, port), timeout=DEFAULT_TIMEOUT)
+    except PySnmpError:
+        # Then try IPv6.
+        try:
+            target = Udp6TransportTarget((host, port), timeout=DEFAULT_TIMEOUT)
+        except PySnmpError as err:
+            _LOGGER.error("Invalid SNMP host: %s", err)
+            return
```

Comment:
Should we check which type of address it is first and try the right transport?

https://docs.python.org/3/library/ipaddress.html#ipaddress.ip_address

Reply:
@bdraco I first wanted to do that as that'd be far cleaner than the try and try again on failure approach, but the issue is that `host` can be a DNS name here with the `UdpTransportTarget` and `Udp6TransportTarget` functions handling the getaddrinfo calls.

So if we wanted to do this, we'd need to move the getaddrinfo calls into sensor.py, then iterate over the results, figure out the family and based on family call the correct transport, then on failure, go over it again with the next result. This would be needed to properly handled DNS records, especially records that can have either multiple values or be dual-stack (both A and AAAA).

Honestly at that point, I think I'd prefer to understand why pysnmp doesn't offer a dual-stack transport implementation as it really feels like something that should be handled at that level instead.

Reply:
The home assistant docs for the `host` field does talk about `IP address` so I also thought we may get away with just not caring about DNS, but reality is that it does work currently and the default value for `host` in the docs is `localhost` which isn't actually an IP address and does rely on address resolution...

Reply:
Not as clean as I would have hoped, but that makes sense to me. 👍